litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -2139,47 +2121,42 @@ private LayoutStateFuture(
                           source,
                           extraAttribution);
                   synchronized (LayoutStateFuture.this) {
-                    if (released) {
-                      result.releaseRef();
-                      return null;
-                    } else {
-                      layoutState = result;
-                      return result;
-                    }
+                    layoutState = result;
```

Comment:
good catch, releasing on the old layoutState makes sense here.

Reply:
actually, since this is inside the callable which is only ever called once, there is no "old layoutState"

But the layoutState that is set here is released when the LayoutStateFuture's refCount reaches 0.

Reply:
I meant we can just not release here in the case that the future was released (which is the only case we return null). In fact, by inspection I think the only way we can return null is if the ComponentTree itself is released from another thread while we're calculating the future.

In that case, we could actually do one of two things: 1) Remove the "if released, return null" logic from the FutureTask. Always return a layoutstate. We won't recycle the LayoutState, but we don't currently do that anyway without futures enabled if the tree was released before computation finished. 2) Keep the logic as it currently is, but null-check in `measure` and `calculateLayout`, assert that we're released (to make sure we aren't missing a case), and exit early since the computation no longer matters.

Reply:
With the current implementation, I don't think the layout state can ever be released from FutureTask#call. The ref count is always decreased only after runAndGet finishes and that's when the LayoutState is also released if needed. I think the only thing we need to make calculateLayoutState be non-nullable is to remove the release logic from the Future task as Andy suggested. @vinc3m1  what do you think?